### PR TITLE
UCP/RNDV: Enabled non-contig datatypes in AM rendezvous.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Enabled non-contig datatypes in AM rendezvous.

## Why ?
The changes are part of UCP transition to new infrastructure.

### CODE REVIEW

File: src/ucp/rndv/proto_rndv.inl [hunk 1]
```diff
@@ -45,13 +45,16 @@ ucp_proto_rndv_rts_request_init(ucp_request_t *req)
...
     }
 
-    status = ucp_datatype_iter_mem_reg(ep->worker->context,
-                                       &req->send.state.dt_iter, rpriv->md_map,
-                                       UCT_MD_MEM_ACCESS_RMA |
-                                       UCT_MD_MEM_FLAG_HIDE_ERRORS,
-                                       UCS_BIT(UCP_DATATYPE_CONTIG));
-    if (status != UCS_OK) {
-        return status;
+    if (!UCP_DT_IS_GENERIC(req->send.state.dt_iter.dt_class)) {
```
Comment (yosefe) [inline src/ucp/rndv/proto_rndv.inl:48]: since ucp_datatype_iter_mem_reg already checks the dt_class, maybe we can pass some parameter to it whether to print error or not? or just not fail if UCT_MD_MEM_FLAG_HIDE_ERRORS is set?

File: src/ucp/am/rndv.c [hunk 2]
```diff
@@ -36,7 +36,8 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_am_rndv_proto_progress, (self),
     size_t max_rts_size;
 
     rpriv  = req->send.proto_config->priv;
-    status = UCS_PROFILE_CALL(ucp_proto_rndv_rts_request_init, req);
+    status = UCS_PROFILE_CALL(ucp_proto_rndv_rts_request_init, req,
+                              UCP_DT_MASK_CONTIG_IOV);
```
Comment (yosefe) [inline src/ucp/am/rndv.c:40]: why not UCP_DT_MASK_ALL?

File: src/ucp/tag/tag_rndv.c [hunk 3]
```diff
@@ -134,7 +134,8 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_tag_rndv_rts_progress, (self),
     rpriv        = req->send.proto_config->priv;
     max_rts_size = sizeof(ucp_rndv_rts_hdr_t) + rpriv->packed_rkey_size;
 
-    status = UCS_PROFILE_CALL(ucp_proto_rndv_rts_request_init, req);
+    status = UCS_PROFILE_CALL(ucp_proto_rndv_rts_request_init, req,
+                              UCS_BIT(UCP_DATATYPE_CONTIG));
```
Comment (yosefe) [inline src/ucp/tag/tag_rndv.c:138]: why not UCP_DT_MASK_ALL?

File: src/ucp/dt/datatype_iter.inl [hunk 4]
```diff
@@ -504,6 +504,8 @@ ucp_datatype_iter_mem_reg(ucp_context_h context, ucp_datatype_iter_t *dt_iter,
                             uct_flags, &dt_iter->type.contig.memh);
     } else if (ucp_datatype_iter_is_class(dt_iter, UCP_DATATYPE_IOV, dt_mask)) {
         return ucp_datatype_iter_iov_mem_reg(context, dt_iter, md_map, uct_flags);
+    } else if (uct_flags & UCT_MD_MEM_FLAG_HIDE_ERRORS) {
+        return UCS_OK;
```
Comment (brminich) [inline src/ucp/dt/datatype_iter.inl:508]: what test is failed? I guess this flags is intended to hide possible registration errors on the specific MDs, while using incorrect datatype looks like a SW bug which needs to be reported

File: src/ucp/rndv/rndv_am.c [hunk 5]
```diff
@@ -77,9 +77,6 @@ ucp_proto_rndv_am_bcopy_complete(ucp_request_t *req)
     if (req->send.rndv.rkey != NULL) {
         ucp_proto_rndv_rkey_destroy(req);
     }
-    ucp_datatype_iter_mem_dereg(req->send.ep->worker->context,
-                                &req->send.state.dt_iter,
-                                UCS_BIT(UCP_DATATYPE_CONTIG));
```
Comment (brminich) [inline src/ucp/rndv/rndv_am.c:82]: is it unrelated bug fix?

File: src/ucp/dt/datatype_iter.inl (hunk 4 above)
Comment (rakhmets) [reply src/ucp/dt/datatype_iter.inl:508]: We perform memory registration inside `ucp_proto_rndv_rts_request_init`.
Previously I added check for `GENERIC` datatype, and replaced it by check for the flag according to the https://github.com/openucx/ucx/pull/8131#discussion_r851996601

File: src/ucp/rndv/rndv_am.c (hunk 5 above)
Comment (rakhmets) [reply src/ucp/rndv/rndv_am.c:82]: This is a redundant memory deregistration since `bcopy` doesn't perform registration.
This didn't lead to an error for `CONTIG` datatype, but for `IOV`.

**Review by yosefe — APPROVED** [review]

File: src/ucp/dt/datatype_iter.inl (hunk 4 above)
Comment (brminich) [reply src/ucp/dt/datatype_iter.inl:508]: i think checking for `GENERIC` datatype was cleaner. Using `UCT_MD_MEM_FLAG_HIDE_ERRORS` for checking proper datatype looks error prone to me. We could just remove the error from this function

Comment (yosefe) [reply src/ucp/dt/datatype_iter.inl:508]: when this function is called without HIDE_ERRORS it means the caller expects registration to really succeed, IMO if we just remove the error it can hide real issues

Comment (brminich) [reply src/ucp/dt/datatype_iter.inl:508]: but this `HIDE` flag is intended for other type of issues. What if we pass invalid datatype, due to a SW bug together with `HIDE` flag? Imo, hiding such errors is not a good approach

Comment (yosefe) [reply src/ucp/dt/datatype_iter.inl:508]: so maybe add another wrapper inline function , so sometimes we would fail if datatype is invalid, and sometimes don't?

**Review by brminich — APPROVED** [review]

**Review by yosefe — DISMISSED** [review]
